File: 01-Programming-Languages/TypeScript/Algorithms/DynamicProgramming/fibonacci.ts
/*
Problem: Fibonacci Number
Description: Compute the nth Fibonacci number using dynamic programming.

Approach:
- Use bottom-up DP with O(n) iteration storing only last two values

Time Complexity: O(n)
Space Complexity: O(1)

Example:
Input: n = 10
Output: 55
*/

function fibonacci(n: number): number {
  if (n <= 1) return n;
  let prev = 0, curr = 1;
  for (let i = 2; i <= n; i++) {
    const next = prev + curr;
    prev = curr;
    curr = next;
  }
  return curr;
}

console.log('fibonacci(10):', fibonacci(10));
console.log('fibonacci(0):', fibonacci(0));
console.log('fibonacci(1):', fibonacci(1));
console.log('fibonacci(20):', fibonacci(20));

File: 01-Programming-Languages/Go/Algorithms/DynamicProgramming/fibonacci.go
/*
Problem: Fibonacci Number
Calculate the nth Fibonacci number using dynamic programming.

Approach:
- Use bottom-up DP with tabulation
- Store only the last two values to optimize space
- F(0) = 0, F(1) = 1, F(n) = F(n-1) + F(n-2)

Time Complexity: O(n)
Space Complexity: O(1)

Example:
Input: n = 10
Output: 55
*/

package main

import "fmt"

func fibonacci(n int) int {
	if n <= 1 {
		return n
	}
	a, b := 0, 1
	for i := 2; i <= n; i++ {
		a, b = b, a+b
	}
	return b
}

func main() {
	n := 10
	result := fibonacci(n)
	fmt.Printf("Input: n = %d\nOutput: %d\n", n, result)
}
